xarray/core/indexing.py

```diff
@@ -645,13 +682,237 @@ def _outer_to_numpy_indexer(key, shape):
     tuple
         Tuple suitable for use to index a NumPy array.
     """
-    if len([k for k in key if not isinstance(k, slice)]) <= 1:
+    if len([k for k in key.tuple if not isinstance(k, slice)]) <= 1:
         # If there is only one vector and all others are slice,
         # it can be safely used in mixed basic/advanced indexing.
         # Boolean index should already be converted to integer array.
-        return tuple(key)
+        return key.tuple
+    else:
+        return _outer_to_vectorized_indexer(key, shape).tuple
+
+
+def _combine_indexers(old_key, old_shape, new_key):
+    """ Combine two indexers. old_key should be a VectorizedIndexer consisting
+    of all np.ndarray, while new_key can be all the indxer types."""
+    if isinstance(new_key, VectorizedIndexer):
+        new_key = _arrayize_vectorized_indexer(new_key, old_shape)
     else:
-        return _outer_to_vectorized_indexer(key, shape)
+        new_key = _outer_to_vectorized_indexer(new_key, old_shape)
+    return VectorizedIndexer(tuple(o[new_key.tuple] for o in
+                                   np.broadcast_arrays(*old_key.tuple)))
+
+
+class IndexingSupport(object):  # could inherit from enum.Enum on Python 3
+    # for backends that support only basic indexer
+    BASIC = 'BASIC'
+    # for backends that support basic / outer indexer
+    OUTER = 'OUTER'
+    # for backends that support outer indexer including at most 1 vector.
+    OUTER_1VECTOR = 'OUTER_1VECTOR'
+    # for backends that support full vectorized indexer.
+    VECTORIZED = 'VECTORIZED'
+
+
+def decompose_indexer(indexer, shape, indexing_support):
+    if isinstance(indexer, VectorizedIndexer):
+        return _decompose_vectorized_indexer(indexer, shape, indexing_support)
+    if isinstance(indexer, OuterIndexer):
+        return _decompose_outer_indexer(indexer, shape, indexing_support)
+    if isinstance(indexer, BasicIndexer):
+        return indexer, BasicIndexer(())
+    raise TypeError('unexpected key type: {}'.format(indexer))
+
+
+def _decompose_vectorized_indexer(indexer, shape, indexing_support):
+    """
+    Decompose vectorized indexer to the successive two indexers, where the
+    first indexer will be used to index backend arrays, while the second one
+    is used to index loaded on-memory np.ndarray.
+
+    Parameters
+    ----------
+    indexer: VectorizedIndexer
+    indexing_support: one of IndexerSupport entries
+
+    Returns
+    -------
+    backend_indexer: OuterIndexer or BasicIndexer
+    np_indexers: a sequence of vectorizedIndexer
```

Comment:
I slightly changed a direction here.
Previously, this function returns a sequence of `VectorizedIndexer`s for in-memory np.ndarray, but now it just returns a single `VectorizedIndexer` by combining multiple `VectorizedIndexer`s into one. It may slightly improve the indexing efficiency.